htsohm/binning.py

```diff
@@ -36,44 +13,32 @@ def select_parents(run_id, children_per_generation, generation):
-    bins = config["number-of-bins"]
-    counts = count_all(run_id)
-    weights = np.zeros([bins, bins, bins])
-    for i in range(bins):
-        for j in range(bins):
-            for k in range(bins):
-                if counts[i,j,k] != 0.:
-                    weights[i,j,k] = counts.sum() / counts[i,j,k]
-    weights = weights / weights.sum()
-
-    ############################################################################
-    # The 3-dimensional list of weights is converted into a 1-dimensional list,
-    # and another 1-dimensional list of material-IDs is also generated.
-    weight_list = []
-    id_list = []
-    for i in range(bins):
-        for j in range(bins):
-            weight_list = np.concatenate([weight_list, weights[i,j,:]])
-            for k in range(bins):
-                bin_ids = []
-                materials = session.query(Material).filter(Material.run_id == run_id, Material.methane_loading_bin == i,
-                    Material.surface_area_bin == j, Material.void_fraction_bin == k,
-                    Material.dummy_test_result.in_(['none', 'pass'])).all()
-                for material in materials:
-                    bin_ids.append(material.id)
-                id_list = id_list + [bin_ids]
+    is selected, a parent is randomly-selected from those materials within that bin.
+    """
+    # Each bin is counted...
+    bins_and_counts = session.query(func.count(Material.id), Material.methane_loading_bin,
+        Material.surface_area_bin, Material.void_fraction_bin).filter(
+        Material.run_id == run_id).group_by(Material.methane_loading_bin,
+        Material.surface_area_bin, Material.void_fraction_bin).all()[1:]
+    bins = [{"ML" : i[1], "SA" : i[2], "VF" : i[3]} for i in bins_and_counts]
+    total = sum([i[0] for i in bins_and_counts])
+    # ...then assigned a weight.
+    weights = [i[0] / float(total) for i in bins_and_counts]
 
     ############################################################################
     # A parent-material is selected for each material in the next generation.
-    next_generation = session.query(Material).filter(Material.run_id == run_id, Material.generation == generation).all()
-
-    for material in next_generation:
-        parents_list = np.random.choice(id_list, p=weight_list)  # weighted sampling function
-        parent_id = np.random.choice(parents_list)
-        material.parent_id = str(parent_id)
+    next_generation = session.query(Material).filter(Material.run_id == run_id,
+        Material.generation == generation).all()
+
+    for child in next_generation:
+        # First, the bin is selected...
+        parent_bin = np.random.choice(bins, p=weights)
+        potential_parents = [i[0] for i in session.query(Material.id).filter(
+            Material.run_id == run_id, Material.methane_loading_bin == parent_bin["ML"],
+            Material.surface_area_bin == parent_bin["SA"],
+            Material.void_fraction_bin == parent_bin["VF"]).all()]
```

Comment:
I would break the result set out from this, i.e. assign the session.query results one line above so that your list comprehension all fits in one line. Generally, a list comprehension should fit in one line, or there is likely a clearer way of doing something.